packages/xsnap/src/build.js

```diff
@@ -239,6 +251,92 @@ async function main(args, { env, stdout, spawn, fs, os }) {
       },
     );
   }
+};
+
+/**
+ * @param {string[]} args
+ * @param {{
+ *   env: Record<string, string | undefined>,
+ *   stdout: typeof process.stdout,
+ *   spawn: typeof import('child_process').spawn,
+ *   fs: {
+ *     existsSync: typeof import('fs').existsSync,
+ *     rmdirSync: typeof import('fs').rmdirSync,
+ *     readFile: typeof import('fs').promises.readFile,
+ *   },
+ *   os: {
+ *     type: typeof import('os').type,
+ *   }
+ * }} io
+ */
+async function main(args, { env, stdout, spawn, fs, os }) {
+  // I solemnly swear I will do no synchronous work followed by a variable
+  // number turns of the event loop.
+  await null;
+
+  const osType = os.type();
+  const platform = {
+    Linux: 'lin',
+    Darwin: 'mac',
+    // Windows_NT: 'win', // One can dream.
+  }[osType];
+  if (platform === undefined) {
+    throw Error(`xsnap does not support platform ${osType}`);
+  }
+
+  // If this is a working copy of xsnap in a checkout of agoric-sdk, we need to
+  // either clone or update submodules.
+  // Otherwise, we are running from either an extracted npm tarball or a Docker
+  // image.
+  // For the npm install case, we should not attempt to update Git submodules,
+  // and should make the binary from the published source.
+  // For the Docker case, we can expect the binary to have been made and added
+  // by the host, and for the native module sources to have been ommitted,
+  // as per .dockerignore in this repository:
+  //
+  //   packages/xsnap/moddable
+  //   packages/xsnap/xsnap-native
+  //   !packages/xsnap/xsnap-native/xsnap/build/bin
+  //
```

Comment:
If you still need these semantics, I’d suggest doing the xsnap build in a stage of Dockerfile.sdk which has the compiler toolchain (i.e. no host build necessary), and in a later stage:
```
RUN rm -rf packages/xsnap/moddable packages/xsnap/xsnap-native && mkdir -p packages/xsnap/xsnap-native/xsnap/build/bin
COPY --from=<earlier-stage> /usr/src/agoric-sdk/packages/xsnap/xsnap-native/xsnap/build/bin/* packages/xsnap/xsnap-native/xsnap/build/bin/
```

That doesn’t need a `.dockerignore` change.

Reply:
I think this requires a base Docker image that includes the build toolchain, including make, git, and the CC, which are evidently absent.

Is it possible to contrive a mechanism that builds xsnap in one container and shares the build product into the next?

Reply:
michaelfig answers out-of-band: Build stages. We can do this.

https://docs.docker.com/build/building/multi-stage/

Reply:
Evidently, you already did this. I pulled out the `.dockerignore` change since it’s no longer helpful or necessary. I’ve also removed the ruminations on the fictional Docker environment where we want `yarn build` to work for `xsnap` when there’s a binary already present but no sources. I left the code mostly the same since there’s no downside.